source/Fleet.cpp

```diff
@@ -496,45 +531,33 @@ void Fleet::Place(const System &system, Ship &ship)
 int64_t Fleet::Strength() const
 {
 	int64_t sum = 0;
-	for(const Variant &variant : variants)
-	{
-		int64_t thisSum = 0;
-		for(const Ship *ship : variant.ships)
-			thisSum += ship->Cost();
-		sum += thisSum * variant.weight;
-	}
+	for(auto &variant : variants)
+		sum += variant.first.Strength() * variant.second;
+	for(auto &variant : stockVariants)
+		sum += variant.first->Strength() * variant.second;
 	return sum / total;
 }
 
 
 
-Fleet::Variant::Variant(const DataNode &node)
-{
-	weight = 1;
-	if(node.Token(0) == "variant" && node.Size() >= 2)
-		weight = node.Value(1);
-	else if(node.Token(0) == "add" && node.Size() >= 3)
-		weight = node.Value(2);
-	
-	for(const DataNode &child : node)
-	{
-		int n = 1;
-		if(child.Size() >= 2 && child.Value(1) >= 1.)
-			n = child.Value(1);
-		ships.insert(ships.end(), n, GameData::Ships().Get(child.Token(0)));
-	}
-}
-
-
-
-const Fleet::Variant &Fleet::ChooseVariant() const
+const Variant &Fleet::ChooseVariant() const
 {
 	// Pick a random variant based on the weights.
 	unsigned index = 0;
-	for(int choice = Random::Int(total); choice >= variants[index].weight; ++index)
-		choice -= variants[index].weight;
-	
-	return variants[index];
+	// Choose between stockVariants and variants.
+	int chosen = Random::Int(total);
+	if(chosen < stockTotal)
+	{
+		for(int choice = Random::Int(stockTotal); choice >= stockVariants[index].second; ++index)
```

Comment:
Why Random again? Is it not possible to re-use chosen here? (It is in the same range as this random number.) Same for the for-loop below.

Reply:
The first random int call is from 0 to `total`, whereas this one is from 0 to `stockTotal`. `stockTotal` is the size of `stockVariants`, whereas `total` is the size of `stockVariants` + `variants`, so reusing the same chosen random number could result in an out of bounds error.

Reply:
Yes, line 548 gives chosen a range from 0 to `total`.
But the if-statement line 549 makes sure that line 551 only gets executed if the actually chosen number on 548 was in the range of 0 to `stockTotal`